File: engine/hashindex/compact.go
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

package hashindex

import "fmt"

// ShardStats reports arena usage for one hash shard.
type ShardStats struct {
	ShardIndex uint32
	SlotCount  uint32
	KeyCount   uint32
	ArenaUsed  uint64
	LiveBytes  uint64
}

// IndexStats aggregates per-shard arena usage.
type IndexStats struct {
	Shards [numShards]ShardStats
}

// Stats returns current arena usage for every shard.
func (idx *Index) Stats() IndexStats {
	var out IndexStats
	for i := 0; i < numShards; i++ {
		seg := idx.shards[i]
		if seg == nil {
			continue
		}
		out.Shards[i] = seg.stats(uint32(i))
	}
	return out
}

func (s *shard) stats(shardIndex uint32) ShardStats {
	s.mu.RLock()
	defer s.mu.RUnlock()
	if s.data == nil {
		return ShardStats{ShardIndex: shardIndex}
	}
	st := ShardStats{
		ShardIndex: shardIndex,
		SlotCount:  s.slotCount(),
		KeyCount:   s.keyCount(),
		ArenaUsed:  s.arenaUsed(),
	}
	st.LiveBytes = s.liveBytesLocked(nil)
	return st
}

// VersionFilter returns the version chain to retain for key (newest first).
type VersionFilter func(key []byte, chain []Version) []Version

func (s *shard) liveBytesLocked(filter VersionFilter) uint64 {
	data := s.data
	if data == nil {
		return 0
	}
	table := int(s.tableOff())
	slots := s.slotCount()
	var live uint64
	for slot := uint32(0); slot < slots; slot++ {
		recOff := readU64(data, table+int(slot)*8)
		if recOff == 0 {
			continue
		}
		key := s.readKey(recOff)
		chain := s.readChainLocked(recOff)
		versions := chain
		if filter != nil {
			versions = filter(key, chain)
		}
		if len(versions) == 0 {
			continue
		}
		live += uint64(12 + len(key))
		live += uint64(len(versions)) * versionNodeSz
	}
	return live
}

func (s *shard) readChainLocked(recOff uint64) []Version {
	data := s.data
	var chain []Version
	for node := s.versionHead(recOff); node != 0; node = readU64(data, int(node)+versionSize) {
		if int(node)+versionNodeSz > len(data) {
			break
		}
		chain = append(chain, readVersion(data, int(node)))
	}
	return chain
}

type compactResult struct {
	ArenaBefore uint64
	ArenaAfter  uint64
	KeysBefore  uint32
	KeysAfter   uint32
}

// CompactShard copies live entries into a fresh arena for shard i.
func (idx *Index) CompactShard(shardIndex int, filter VersionFilter) (ShardStats, error) {
	if shardIndex < 0 || shardIndex >= numShards {
		return ShardStats{}, fmt.Errorf("hashindex: invalid shard index %d", shardIndex)
	}
	seg := idx.shards[shardIndex]
	if seg == nil {
		return ShardStats{}, fmt.Errorf("hashindex: shard %d is nil", shardIndex)
	}
	res, err := seg.compact(filter)
	if err != nil {
		return ShardStats{}, err
	}
	return ShardStats{
		ShardIndex: uint32(shardIndex),
		SlotCount:  seg.slotCount(),
		KeyCount:   seg.keyCount(),
		ArenaUsed:  res.ArenaAfter,
		LiveBytes:  res.ArenaAfter,
	}, nil
}

// CompactAll rewrites every shard that has keys, using filter to trim version chains.
func (idx *Index) CompactAll(filter VersionFilter) (IndexStats, error) {
	var out IndexStats
	for i := 0; i < numShards; i++ {
		seg := idx.shards[i]
		if seg == nil {
			continue
		}
		if _, err := seg.compact(filter); err != nil {
			return out, err
		}
		out.Shards[i] = seg.stats(uint32(i))
	}
	return out, nil
}

func (s *shard) compact(filter VersionFilter) (compactResult, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.data == nil {
		return compactResult{}, fmt.Errorf("hashindex: shard is closed")
	}

	res := compactResult{
		ArenaBefore: s.arenaUsed(),
		KeysBefore:  s.keyCount(),
	}

	type keyEntry struct {
		key      []byte
		versions []Version
	}

	data := s.data
	slots := s.slotCount()
	table := int(s.tableOff())
	var entries []keyEntry

	for slot := uint32(0); slot < slots; slot++ {
		recOff := readU64(data, table+int(slot)*8)
		if recOff == 0 {
			continue
		}
		key := s.readKey(recOff)
		chain := s.readChainLocked(recOff)
		versions := chain
		if filter != nil {
			versions = filter(key, chain)
		}
		if len(versions) == 0 {
			continue
		}
		entries = append(entries, keyEntry{key: key, versions: versions})
	}

	slotCount := slots
	if slotCount == 0 {
		slotCount = initialSlots
	}
	tableBytes := uint64(slotCount) * 8
	var liveBytes uint64
	for _, e := range entries {
		liveBytes += uint64(12+len(e.key)) + uint64(len(e.versions))*versionNodeSz
	}
	newSize := int64(headerSize) + int64(tableBytes) + int64(liveBytes) + headerSize
	if newSize < int64(headerSize)+int64(tableBytes)+headerSize {
		newSize = int64(headerSize) + int64(tableBytes) + headerSize
	}

	newData := make([]byte, newSize)
	writeU64(newData, hdrMagicOff, magic)
	writeU32(newData, hdrVersionOff, formatVersion)
	writeU32(newData, hdrSlotCountOff, slotCount)
	writeU32(newData, hdrKeyCountOff, 0)
	tableOff := uint64(headerSize)
	arenaOff := tableOff + tableBytes
	writeU64(newData, hdrTableOffOff, tableOff)
	writeU64(newData, hdrArenaOffOff, arenaOff)
	writeU64(newData, hdrArenaUsedOff, 0)

	newShard := &shard{data: newData}
	newShard.setArenaUsed(0)

	for _, e := range entries {
		recSize := 12 + len(e.key)
		recOff, err := newShard.alloc(recSize)
		if err != nil {
			return res, err
		}
		writeU32(newShard.data, int(recOff), uint32(len(e.key)))
		writeU64(newShard.data, int(recOff)+4, 0)
		copy(newShard.data[int(recOff)+12:], e.key)

		var head uint64
		for i := len(e.versions) - 1; i >= 0; i-- {
			nodeOff, err := newShard.alloc(versionNodeSz)
			if err != nil {
				return res, err
			}
			writeVersion(newShard.data, int(nodeOff), e.versions[i])
			writeU64(newShard.data, int(nodeOff)+versionSize, head)
			head = nodeOff
		}
		newShard.setVersionHead(recOff, head)

		if err := newShard.insertKeySlot(e.key, recOff); err != nil {
			return res, err
		}
	}

	s.data = newShard.data
	res.ArenaAfter = s.arenaUsed()
	res.KeysAfter = s.keyCount()
	return res, nil
}

File: rust/crates/turnstone-hashindex/src/compact.rs
// Copyright (c) 2026 Kiruba Sankar Swaminathan
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root of this source tree.

use std::collections::HashMap;
use std::sync::Arc;

use crate::index::Index;
use crate::shard::{live_bytes_for_map, Shard};
use crate::Version;

/// Per-shard arena usage report.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct ShardStats {
    pub shard_index: u32,
    pub slot_count: u32,
    pub key_count: u32,
    pub arena_used: u64,
    pub allocated_bytes: u64,
    pub live_bytes: u64,
}

/// Aggregated per-shard arena usage.
#[derive(Debug, Clone)]
pub struct IndexStats {
    pub shards: [ShardStats; crate::encoding::NUM_SHARDS],
}

impl Default for IndexStats {
    fn default() -> Self {
        Self {
            shards: [ShardStats::default(); crate::encoding::NUM_SHARDS],
        }
    }
}

/// Returns the version chain to retain for a key (newest first).
pub type VersionFilter<'a> = dyn Fn(&[u8], &[Version]) -> Vec<Version> + 'a;

struct RestoreEnforce(Arc<crate::index::IndexInner>);

impl Drop for RestoreEnforce {
    fn drop(&mut self) {
        self.0.set_enforce_limit(true);
    }
}

impl Index {
    pub fn stats(&self) -> IndexStats {
        let mut out = IndexStats::default();
        for (i, seg) in self.shards.iter().enumerate() {
            out.shards[i] = seg.stats(i as u32);
        }
        out
    }

    pub fn filtered_live_bytes(
        &self,
        shard_index: i32,
        filter: Option<&VersionFilter<'_>>,
    ) -> (u64, u64, u32) {
        if shard_index < 0 || shard_index as usize >= crate::encoding::NUM_SHARDS {
            return (0, 0, 0);
        }
        self.shards[shard_index as usize].filtered_live_bytes(filter)
    }

    pub fn compact_shard(
        &self,
        shard_index: i32,
        filter: Option<&VersionFilter<'_>>,
    ) -> Result<ShardStats, Box<dyn std::error::Error + Send + Sync>> {
        if shard_index < 0 || shard_index as usize >= crate::encoding::NUM_SHARDS {
            return Err(format!("hashindex: invalid shard index {shard_index}").into());
        }
        let seg = &self.shards[shard_index as usize];
        seg.compact(filter)?;
        Ok(seg.stats(shard_index as u32))
    }

    pub fn compact_all(
        &self,
        filter: Option<&VersionFilter<'_>>,
    ) -> Result<IndexStats, Box<dyn std::error::Error + Send + Sync>> {
        let mut out = IndexStats::default();
        for (i, seg) in self.shards.iter().enumerate() {
            seg.compact(filter)?;
            out.shards[i] = seg.stats(i as u32);
        }
        Ok(out)
    }
}

pub(crate) fn compact_shard(
    shard: &Shard,
    filter: Option<&VersionFilter<'_>>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    shard.parent().set_enforce_limit(false);
    let _restore = RestoreEnforce(Arc::clone(shard.parent()));

    let mut st = shard.state.write();
    if st.closed {
        return Err("hashindex: shard is closed".into());
    }

    let arena_before = live_bytes_for_map(&st.keys);
    let mut next: HashMap<Vec<u8>, Vec<Version>> = HashMap::new();
    for (key, chain) in st.keys.drain() {
        let versions = if let Some(f) = filter {
            f(&key, &chain)
        } else {
            chain
        };
        if !versions.is_empty() {
            next.insert(key, versions);
        }
    }

    let before_alloc = crate::encoding::HEADER_SIZE as i64 + arena_before as i64;
    st.keys = next;
    let arena_after = live_bytes_for_map(&st.keys);
    let after_alloc = crate::encoding::HEADER_SIZE as i64 + arena_after as i64;
    drop(st);

    let delta = after_alloc - before_alloc;
    if delta != 0 {
        shard
            .parent()
            .account_delta(delta)
            .map_err(|e| -> Box<dyn std::error::Error + Send + Sync> { Box::new(e) })?;
    }

    Ok(())
}
